chapter1/poisson.ipynb

```diff
@@ -273,9 +277,120 @@
     "# print the result\n",
     "print(h1_error)"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "14d92a71-8c0f-4a91-8f56-c405f3fc76d1",
+   "metadata": {},
+   "source": [
+    "### Visualization\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "968a2165-ab75-4bf0-b943-daab3bd9fa2d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
+    "\n",
+    "def refine_array_1d(breaks, N):\n",
+    "    \"\"\"Refine a 1D array by inserting N points between each pair of original values.\"\"\"\n",
+    "    refined = np.concatenate([np.linspace(breaks[i], breaks[i+1], N+1)[:-1] for i in range(len(breaks)-1)])\n",
+    "    return np.append(refined, breaks[-1])  # Add the last point\n",
+    "\n",
+    "def plot_solutions(Vh, uh, ue, N=10):\n",
+    "    \"\"\"\n",
+    "    Refine the grid, evaluate solutions, compute errors, and plot results.\n",
+    "\n",
+    "    Parameters:\n",
+    "        Vh: The finite element space (must have `spaces` attribute with `breaks`).\n",
+    "        uh: The numerical solution function.\n",
+    "        ue: The exact solution function.\n",
+    "        N: Number of points to insert between breaks (default: 10).\n",
+    "    \"\"\"\n",
+    "    # Generate refined grid for visualization\n",
+    "    eta1 = refine_array_1d(Vh.spaces[0].breaks, N)\n",
+    "    eta2 = refine_array_1d(Vh.spaces[1].breaks, N)\n",
+    "\n",
+    "    # Evaluate numerical solution on the refined grid\n",
+    "    num = np.array([[uh(e1, e2) for e2 in eta2] for e1 in eta1])\n",
+    "\n",
+    "    # Compute exact solution\n",
+    "    ex = np.array([[phi_exact(e1, e2) for e2 in eta2] for e1 in eta1])\n",
```

Comment:
it seems to me that this line triggers an error when the notebook is executed. @anushkasinghh did you verify that you could run the notebooks ?

When the CI tests are run with the PR (probably after each new commit) you can also check that the "book" (which will be deployed after this PR is merged) looks as it should, by downloading the artifacts as described in the [readme](https://github.com/pyccel/IGA-Python?tab=readme-ov-file#contributing).

But you should first check that you can run the notebook correctly before committing, that will be simpler :)